app/lib/package/models.dart

```diff
@@ -340,27 +340,31 @@ class Package extends db.ExpandoModel<String> {
     PackageVersion pv, {
     required Version dartSdkVersion,
     required Version flutterSdkVersion,
+    bool existingLatestIsRetracted = false,
   }) {
```

Comment:
I think we should seriously consider being more functional here.

Instead of looking at:
 * `Package` entity,
 * `PackageVersion` entity that was just published, and,
 * SDK versions.
And then try to decide of `pv` should replace any of the current designations as latest, preview, prerelease in the `Package` entity.

----

We should instead:
 * Fetch all `PackageVersion` entities for the given package.
 * Determine latest, preview and prerelease using:
    * the SDK versions
    * All `PackageVersion` entities for the given package.
 * Update the `PackageEntity` to the newly determined values for latest, preview and prerelease.

In particular, when determining what `latestVersionKey` should be, **we should never read** `PackageEntity.latestVersionKey`.

The value of `PackageEntity.latestVersionKey` should be derived from:
 * the SDK versions,
 * All `PackageVersion` entities for the given package.

That removes a LOT of corner cases, and it becomes much simpler to keep the code correct.

Yes, it's more expensive to load all `PackageVersion` entities for a given package. But we rarely change a property that causes us to recompute `PackageEntity.latestVersionKey` (and friends). This mostly happens when:
 * a package is published,
 * an admin action flips retracted state on a version, or,
 * a new SDK version is published.

All of these are rare events.

Reply:
Added a tracking issue, merging the fix for now.